core/src/main/java/io/opencensus/stats/export/StatsExporter.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2016-17, OpenCensus Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opencensus.stats.export;
+
+import io.opencensus.stats.ViewData;
+import java.util.Collection;
+import javax.annotation.concurrent.ThreadSafe;
+
+/** A service that is used by the library to export {@code ViewData}. */
+@ThreadSafe
+public abstract class StatsExporter {
+
+  private static final NoopStatsExporter NOOP_STATS_EXPORTER = new NoopStatsExporter();
+
+  /**
+   * Returns the no-op implementation of the {@code StatsExporter}.
+   *
+   * @return the no-op implementation of the {@code StatsExporter}.
+   */
+  public static NoopStatsExporter getNoopStatsExporter() {
+    return NOOP_STATS_EXPORTER;
+  }
+
+  /**
+   * Registers a new service handler that is used by the library to export {@code ViewData}.
+   *
+   * @param name the name of the service handler. Must be unique for each service.
+   * @param handler the service handler that is called for each ended sampled span.
+   */
+  public abstract void registerHandler(String name, Handler handler);
+
+  /**
+   * Unregisters the service handler with the provided name.
+   *
+   * @param name the name of the service handler that will be unregistered.
+   */
+  public abstract void unregisterHandler(String name);
+
+  /**
+   * An abstract class that exports recorded {@code ViewData} in their own format.
+   *
+   * <p>To export data this MUST be register to to the ExportComponent using {@link
+   * #registerHandler(String, Handler)}.
+   */
+  public abstract static class Handler {
+
+    /**
+     * Exports a list of {@code ViewData}.
+     *
+     * <p>Implementation SHOULD not block the calling thread. It should execute the export on a
+     * different thread if possible.
+     *
+     * @param viewDataList a list of {@code ViewData} objects to be exported.
+     */
+    public abstract void export(Collection<ViewData> viewDataList);
```

Comment:
I don't know how this interacts with ViewManager and stuff. Can you give some details?

Reply:
In `StatsManager` we have a `StatsEvent` class, which implements `EventQueue.Entry` and records stats: https://github.com/census-instrumentation/opencensus-java/blob/master/core_impl/src/main/java/io/opencensus/implcore/stats/StatsManager.java#L78. In the `process()` method, after stats are recorded, a `ViewData` will be added by `StatsExporter.addViewData()`. Similar to span end process: https://github.com/census-instrumentation/opencensus-java/blob/master/impl_core/src/main/java/io/opencensus/implcore/trace/StartEndHandlerImpl.java#L117

Reply:
I am not asking about implementation details. I think the design that @dinooliva told me he wants to have a per-view registration. Does this register for all the views? how does this interact with the per-view registration?

Reply:
We haven't talked about this yet. By now export is enabled for all views by default and doesn't interact with view registrations. Will discuss with @dinooliva later.

Reply:
If that is the case the API looks good.

Reply:
> By now export is enabled for all views by default and doesn't interact with view registrations.

Does that mean that the set of views used by StatsExporter is completely independent of the views registered with the ViewManager?  I thought that ViewManager would also handle pushing data.  Then we could have a separate library for exporting to StackDriver that registers StackDriver-specific handlers with the ViewManager.

Reply:
> Does that mean that the set of views used by StatsExporter is completely independent of the views registered with the ViewManager?

No it's not independent, only stats of the views registered by Viewmanager will be exported.

> I thought that ViewManager would also handle pushing data.

My idea is pushing stats will be a implementation detail: whenever a record() [StatsEvent](https://github.com/census-instrumentation/opencensus-java/blob/master/core_impl/src/main/java/io/opencensus/implcore/stats/StatsManager.java#L64) happens, `StatsManager` will invoke some method of StatsExporter and push ViewData to it.

> we could have a separate library for exporting to StackDriver that registers StackDriver-specific handlers with the ViewManager.

StackDriver handlers will be in another library (under [exporters](https://github.com/census-instrumentation/opencensus-java/tree/master/exporters)). Are you suggesting we should also move StatsExporter to another library?